### ISSUE TITLE
Error: spawnSync /bin/sh ENOBUFS

### ISSUE THREAD

**Issue by bohendo** [EthWorks/Waffle#179]
I'm trying to consolidate a couple related contract repos into one module. Compiling gives me the following error:

``` 
> waffle waffle.json

Error: spawnSync /bin/sh ENOBUFS
    at Object.spawnSync (internal/child_process.js:1041:20)
    at spawnSync (child_process.js:607:24)
    at Object.execSync (child_process.js:652:15)
    at /root/node_modules/ethereum-waffle/dist/compiler/compileNative.js:48:66
    at step (/root/node_modules/ethereum-waffle/dist/compiler/compileNative.js:32:23)
    at Object.next (/root/node_modules/ethereum-waffle/dist/compiler/compileNative.js:13:53)
    at /root/node_modules/ethereum-waffle/dist/compiler/compileNative.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/root/node_modules/ethereum-waffle/dist/compiler/compileNative.js:3:12)
    at compile (/root/node_modules/ethereum-waffle/dist/compiler/compileNative.js:43:16) {
  errno: 'ENOBUFS',
  code: 'ENOBUFS',
  syscall: 'spawnSync /bin/sh',
  path: '/bin/sh',
  spawnargs: [
    '-c',
    'solc --standard-json --allow-paths /root/modules/contracts/contracts,/root/node_modules'
  ],
  error: [Circular],
  status: null,
  signal: 'SIGTERM',
  output: [
    null,
    <Buffer 7b 22 63 6f 6e 74 72 61 63 74 73 22 3a 7b 22 40 63 6f 6e 6e 65 78 74 2f 63 66 2d 61 64 6a 75 64 69 63 61 74 6f 72 2d 63 6f 6e 74 72 61 63 74 73 2f 63 ... 1114062 more bytes>,
    <Buffer >
  ],
  pid: 30,
  stdout: <Buffer 7b 22 63 6f 6e 74 72 61 63 74 73 22 3a 7b 22 40 63 6f 6e 6e 65 78 74 2f 63 66 2d 61 64 6a 75 64 69 63 61 74 6f 72 2d 63 6f 6e 74 72 61 63 74 73 2f 63 ... 1114062 more bytes>,
  stderr: <Buffer >
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @connext/contracts@1.0.0 build: `waffle waffle.json`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @connext/contracts@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-01-13T12_07_56_024Z-debug.log
```

This is happening w/in a monorepo w dependencies managed by lerna. The monorepo has 4 contract modules that I'm attempting to consolidate into one. waffle.json looks like:

```
{
  "compiler": "native",
  "npmPath": "../../node_modules",
  "legacyOutput": true,
  "compilerOptions": {
    "evmVersion": "constantinople"
  },
  "sourcesPath": "./contracts"
}
```

I'm compiling using an alpine docker image with a native solc executable installed.

I'm suspicious that I'm trying to compile too many contracts all at once and I'm getting some kind of out-of-memory issue (eg trying to create too many buffers).

I'm still investigating the possibility of a circular dependency and I'll update if I find one.

At the moment, a working hack-around is to create two waffle.json files that each compile half the the contracts. The build succeeds if done in two steps like this.

**Comment by bohendo** [comment]
Update: after a careful read-through, I'm confident there aren't any circular dependencies lurking.

To repro:
 - `git clone https://github.com/ConnextProject/indra.git`
 - `git checkout 676-auto-test-contracts`
 - `make contracts`

& the above error should print after installing lots of dependencies (sorry, first time running `make contracts` will take a while but subsequent runs will go a lot faster)

**Comment by bohendo** [comment]
Note, `truffle compile` fails with same error while using the native solc compiler
```
Error: Error: spawnSync /bin/sh ENOBUFS
    at Object.compile (/root/node_modules/truffle/build/webpack:/packages/workflow-compile/legacy/index.js:80:1)
Truffle v5.1.6 (core: 5.1.6)
Node v12.13.0
```
So this error might actually be coming from something in the `ethereum/solc:0.5.15-alpine` image instead of waffle itself..

**Comment by bohendo** [comment]
Closing issue.

Summary:
 - waffle is only the messenger, not the source of this problem.
 - The error seems to arise from the native `solc` compiler in `ethereum/solc:0.5.15-alpine`, I've opened [issue 8133 in the solidity repo](https://github.com/ethereum/solidity/issues/8133) to continue the investigation.
 - `soljs` works fine, that's what I'm going to use for now.

**Comment by sz-piotr** [comment]
A true hero. You found the error, did the proper investigation and documented everything along the way. We need more contributors like you. Thank you.

**Comment by bohendo** [comment]
Thank you for the praise but I've revisited this issue (bc `solcjs` is too slow) and now I see that my previous comment is wrong: waffle is indeed the source of the problem.

I dug in a bit & discovered that the fix is very simple & have submitted PR #205 to resolve this issue for real.

**Comment by msieczko** [comment]
Closed in #212
